### DISCUSSION TITLE
Update local »content://« calendars?

### DISCUSSION THREAD

**Discussion by fhaftmann** [bitfireAT/icsx5#47] [Q&A]
My envisaged use case is to have a calendar stored locally as ical file on an android phone and syncing it directly over MTP.

ICSx⁵ seems to be the nearest app candidate for this scenario.

Would it be architecturally feasible to extend it in such a way?

This is not a plea for implementation, but a request for expertise whether it would be a good idea in principle.

**Comment by rfc2822** [comment]
ICSx⁵ supports the Storage Access Framework (SAF, `content://` URIs). Did you try to use it the way you think?

>This is not a plea for implementation, but a request for expertise whether it would be a good idea in principle.

In my ears, transferring calendars over MTP doesn't sound like a fail-safe solution. Personally, I'd prefer a real server, maybe in the home network or even on a desktop PC or laptop, and exchange data using this server.

**Reply by fhaftmann** [reply]
> ICSx⁵ supports the Storage Access Framework (SAF, `content://` URIs). Did you try to use it the way you think?

Yes, I tried SAF via content-URIs, but those appear to be read-only e. g. in the Etar app.

> > This is not a plea for implementation, but a request for expertise whether it would be a good idea in principle.
> 
> In my ears, transferring calendars over MTP doesn't sound like a fail-safe solution. Personally, I'd prefer a real server, maybe in the home network or even on a desktop PC or laptop, and exchange data using this server.

The MTP approach is surely somehow special and elementary, but in my particular application scenario I want to avoid any server component.

**Reply by devvv4ever** [reply]
ICSx⁵ can't write content. It's only designed to read data. Maybe "iCal Import / Export" is an app that offers this - but I don't know.

**Reply by fhaftmann** [reply]
> ICSx⁵ can't write content. It's only designed to read data.

That is actually the core of my question: I am aware that in the current state of the implementation ICSx⁵ does not write content, but whether an extension would be feasible architecturally. Taking your statement as a maintainer »it’s only designed to read« literally, I would deduce a »no«.

> Maybe "iCal Import / Export" is an app that offers this - but I don't know.

I am aware of this app, but it has not been seriously maintained for years now. Actually, I tried to step in as maintainer but I don’t think I can manage that: there has never been a full hands-over of access resources, there a plenty of problems on different android versions, dependencies are stale, the code has serious flaws, there is no test coverage e. g. for conversions etc. I deem the effort to put that in shape is more intrusive than to switch to a solid base.

ICSx⁵ and the libraries behind at a first glance seem to be well-maintained and well-designed, with the option of commerical support also.
Using that as a starting point, I see two options: extend ICSx⁵ with local write support (what is discussed in this thread), or extend Etar with an import/export facility for whole calendars, maybe switching to ical4android as a base.

If the first option is ruled out, the second remains.
